llama.cpp

```diff
@@ -2232,15 +2232,32 @@ static void llm_load_vocab(
     if (vocab.type == LLAMA_VOCAB_TYPE_SPM) {
         vocab.linefeed_id = llama_byte_to_token(vocab, '\n');
     } else {
-        vocab.linefeed_id = llama_tokenize_internal(vocab, "\u010A", false)[0];
+        const std::vector<int> ids = llama_tokenize_internal(vocab, "\u010A", false);
+        GGML_ASSERT(!ids.empty() && "model vocab missing newline token");
+        vocab.linefeed_id = ids[0];
     }
 
     // special tokens
-    GGUF_GET_KEY(ctx, vocab.special_bos_id, gguf_get_val_u32, GGUF_TYPE_UINT32, false, kv(LLM_KV_TOKENIZER_BOS_ID));
-    GGUF_GET_KEY(ctx, vocab.special_eos_id, gguf_get_val_u32, GGUF_TYPE_UINT32, false, kv(LLM_KV_TOKENIZER_EOS_ID));
-    GGUF_GET_KEY(ctx, vocab.special_unk_id, gguf_get_val_u32, GGUF_TYPE_UINT32, false, kv(LLM_KV_TOKENIZER_UNK_ID));
-    GGUF_GET_KEY(ctx, vocab.special_sep_id, gguf_get_val_u32, GGUF_TYPE_UINT32, false, kv(LLM_KV_TOKENIZER_SEP_ID));
-    GGUF_GET_KEY(ctx, vocab.special_pad_id, gguf_get_val_u32, GGUF_TYPE_UINT32, false, kv(LLM_KV_TOKENIZER_PAD_ID));
+    {
+        const std::vector<std::tuple<enum llm_kv, int32_t &>> special_token_types = {
+            { LLM_KV_TOKENIZER_BOS_ID, vocab.special_bos_id },
+            { LLM_KV_TOKENIZER_EOS_ID, vocab.special_eos_id },
+            { LLM_KV_TOKENIZER_UNK_ID, vocab.special_unk_id },
+            { LLM_KV_TOKENIZER_SEP_ID, vocab.special_sep_id },
+            { LLM_KV_TOKENIZER_PAD_ID, vocab.special_pad_id },
+        };
+        for (const auto & it : special_token_types ) {
+            const std::string key = kv(std::get<0>(it));
+            int32_t & id = std::get<1>(it), old_id = id;
+
+            GGUF_GET_KEY(ctx, id, gguf_get_val_u32, GGUF_TYPE_UINT32, false, key);
+            if (id != -1 && (id < 0 || size_t(id) >= vocab.id_to_token.size())) {
```

Comment:
Would it be clearer to write this as `id < -1 || size_t(id) >= vocab.id_to_token.size()`?

Reply:
I had that same thought and then I realized it can lead to converting `-1` to `size_t`. The way the boolean short circuiting is set up is to avoid that, even though it's a little more complicated.

Reply:
Then `id < -1 || id >= ptrdiff_t(vocab.id_to_token.size())`?

Reply:
What I meant is if you have `id < -1 || size_t(id) >= x` the `id < -1` doesn't stop the expression from evaluating when `id == -1`. So then you do `size_t(-1) >= x`. Isn't `size_t(-1)` going to be a huge number since `size_t` is unsigned?

Reply:
Did you see my comment here? https://github.com/ggerganov/llama.cpp/pull/3635#discussion_r1360781711

Reply:
> Did you see my comment here?

Sort-of. I was still stuck on what you originally said, also I didn't know the properties of `ptrdiff_t` off the time of my head. It doesn't look like that type is used in the project source at all (except for some 3rd party stuff like a JSON lib).

Personally, I'd say the version with `ptrdiff_t` would be harder for me to understand than the original version just because all the information is there to think through the boolean logic but if you don't know how `ptrdiff_t` works you don't know what's correct.

I was thinking there may be a simpler solution, something like: 

```c++
if (id < -1 || size_t(id + 1) > vocab.id_to_token.size())
```

The first part excludes anything below `-1`, then since we add `1`, the valid values are between 0 and the size of the vector inclusive now. Pretty simple, hopefully not too cute/tricky. What do you think?

Reply:
ptrdiff_t is a standard C++ difference type that is large enough to hold the difference between two pointers, positive or negative - it will be 32 bits on 32-bit platforms, 64 bits on 64-bit platforms. It's used in the C++ standard library where a large signed type is needed. There is also intptr_t, although it's not guaranteed to exist.

Maybe I'm just used to python where you can compare two numbers without extra operations to check if they are in range of the data type being used :)

Reply:
> ptrdiff_t is a standard C++ difference type that is large enough to hold the difference between two pointers, positive or negative

Right, and I could guess a lot of that from the name, but this is C++ so all the semantics aren't necessarily going to be intuitive. That's what makes it harder for me to deal with than the original version. Also, if I'm going to be okay with it going into a 32bit signed type, I could just use `int32_t` like the other stuff.

What about the alternative I suggested?

> Maybe I'm just used to python where you can compare two numbers without extra operations to check if they are in range of the data type being used

In this case, you don't really _have_ to, so this is mostly just to stop compiler warnings by explicitly saying "Yes, I meant to do that". I get your point though. I'm by no means a C++ person, I actually never touched it until I started contributing to this project. I'd rather be writing all this in Rust!

Reply:
> Also, if I'm going to be okay with it going into a 32bit signed type, I could just use `int32_t` like the other stuff.

I like ptrdiff_t because it's a signed type where `sizeof(ptrdiff_t) == sizeof(size_t)` (on any platform I know of), just like ssize_t, which is used on *nix for syscalls like read that may need to return `-1`.